docs/reprocess-report.md

```diff
@@ -0,0 +1,122 @@
+- ...
+- update `centrifugation.py`: set up-to-date `fingerprint` table checksum in `HttpRequestFPFeeder.__init__()`. There is no need to bump `CODE_VER` for feature-based fingerprints as we are 100% confident that reprocessing is not needed and ongoing data processing is paused.
+- ...
+- there is no need for reprocessing as there is no possibility for false negative here
+
+_A temporary table_ is not necessary a result of `CREATE TEMPORARY TABLE`, it may also be a query executed on a read-only replica with faster disk drives with the output of the query directed to a local file that becomes `UNLOGGED` table on a master via out-of-band data transfer or via _Foreign Data Wrapper_.
+
+Unfortunately, it's not trivial to give a concrete example of the queries as these examples have to be kept in-sync with the rest of the code and, what's more important, different cardinality of the tables may need different strategies for UPDATE. E.g. [CREATE TABLE + rename](https://github.com/ooni/pipeline/pull/144#issuecomment-483365330) strategy may be order of magnitude more performant than `UPDATE` when the UPDATE touches _many_ rows (it was touching ≈5% of rows in the case).
+
+## Case: new feature table
+
+The rule of thumb is: if you are not going to use the extracted features for search or aggregation, you should rather consider leaving JSON as-is without bloating the MetaDB. Maintained table should be an asset, not just a liability of maintenance for the sake of maintenance.
+
+One may want to use [commit adding `vanilla_tor` stats](https://github.com/ooni/pipeline/commit/902e6751340dd515096214f74c739751c9ddca55) for inspiration, but the code evolved a bit since than.
+
+- Add new feature table. Avoid foreign keys, those are very slow to verify during batch ingestion (as of PostgreSQL 9.6).
+- Bump `CODE_VER`, set `min_compat_code_ver` for the new feeder
+- `TheFeeder.row()` creates a string that is suitable for sending to the table via `COPY`
+- `TheFeeder.pop()` removes fields from the JSON object those are completely ingested by the feeder and should NOT be considered a part of the _residual_
+- test, deploy, reprocess all (or the affected) buckets under GNU Make control
+
+One may save significant amount of CPU time marking old _autoclaved_ files as already processed by the new version of code bumping their corresponding `code_ver` in the database. It may be useful in a case when a feature has to be extracted **only(!)** from a known subset of reports, so the reports that have no data on the specific feature may be skipped safely. Example is extracting a feature of a "low-volume" test. E.g. `web_connectivity` test takes 99.4% of data volume of 2019Q1, so _any_ other test is a low-volume one. Another example is a extracting a feature that was shipped as a part of some specific `software` version, so `autoclaved` having no records coming from the new software may be manually labeled with a newer `code_ver` and skipped safely.
```

Comment:
Can you provide a concrete code example of how to do this?

For example suppose I want to apply this CPU saving to deploy/rollout https://github.com/ooni/pipeline/pull/177

Reply:
#177 touches only `test_name = 'telegram'` and we don't need to reprocess the files those are already current, so it's something close to
```sql
UPDATE autoclaved SET code_ver = 6 WHERE code_ver = 5 AND autoclaved_no IN (
    SELECT autoclaved_no FROM autoclaved WHERE autoclaved_no NOT IN (
        SELECT DISTINCT autoclaved_no FROM report WHERE test_name = 'telegram'));
```

Reply:
Thanks. I added this as a concrete example in the text.